README.md
Florida COVID-19 Data
================
2020-09-08 12:13:58

## Today

| When        | Day        | Positive | New Positive Since | Deaths | New Deaths Since | Total     |
| :---------- | :--------- | :------- | :----------------- | :----- | :--------------- | :-------- |
| Today       | 2020-09-08 | 650,092  | 0                  | 12,067 | 0                | 4,816,873 |
| Yesterday   | 2020-09-07 | 648,269  | 1,823              | 12,023 | 44               | 4,801,684 |
| Last Week   | 2020-09-01 | 631,040  | 19,052             | 11,521 | 546              | 4,675,866 |
| 2 Weeks Ago | 2020-08-25 | 605,502  | 44,590             | 10,717 | 1,350            | 4,466,524 |

Parsed from the [Florida’s COVID-19 Data and Surveillance
Dashboard](https://fdoh.maps.arcgis.com/apps/opsdashboard/index.html#/8d0de33f260d444c852a615dc7837c86).

Prior to 2020-04-25, Florida DOH published updated data twice per day.
Starting 2020-04-25, Florida DOH updates once per day at approximately
11am. Since 2020-05-06, the updates are occasionally later in the day.
In the following charts, for a given day, I use the last value reported
prior to 8am of the subsequent day. I take snapshots of the various FL
DOH data sources at 7am, noon and 7pm daily.

## Confirmed Cases

![](plots/covid-19-florida-daily-test-changes.png)

![](plots/covid-19-florida-deaths-by-day.png)

![](plots/covid-19-florida-county-top-6.png)

<div class="columns">

<div class="column is-full-mobile">

![](plots/covid-19-florida-testing.png)

</div>

<div class="column is-full-mobile">

![](plots/covid-19-florida-total-positive.png)

</div>

</div>

## Hospital and ICU Utilization

| County       | Current COVID Hospitalizations | Change Since Yesterday                  | Available Hospital Beds                      | Available ICU Beds                         |
| :----------- | -----------------------------: | :-------------------------------------- | :------------------------------------------- | :----------------------------------------- |
| All          |                           3163 | <span style="color: #EC4E20">↑ 6</span> | 17552<span style="color: #aaa">/42658</span> | 1544<span style="color: #aaa">/4410</span> |
| Miami-Dade   |                            535 | <span style="color: #EC4E20">↑ 6</span> | 2564<span style="color: #aaa">/5866</span>   | 190<span style="color: #aaa">/741</span>   |
| Broward      |                            345 |                                         | 1189<span style="color: #aaa">/3925</span>   | 100<span style="color: #aaa">/363</span>   |
| Duval        |                            199 |                                         | 1236<span style="color: #aaa">/2551</span>   | 154<span style="color: #aaa">/283</span>   |
| Hillsborough |                            174 |                                         | 833<span style="color: #aaa">/2858</span>    | 82<span style="color: #aaa">/284</span>    |
| Palm Beach   |                            171 |                                         | 1510<span style="color: #aaa">/2544</span>   | 146<span style="color: #aaa">/251</span>   |
| Orange       |                            149 |                                         | 1251<span style="color: #aaa">/3084</span>   | 119<span style="color: #aaa">/253</span>   |
| Pinellas     |                            131 |                                         | 1041<span style="color: #aaa">/2093</span>   | 79<span style="color: #aaa">/219</span>    |
| Gadsden      |                            124 |                                         | 208<span style="color: #aaa">/745</span>     | 0<span style="color: #aaa">/0</span>       |
| Polk         |                            110 |                                         | 451<span style="color: #aaa">/1158</span>    | 32<span style="color: #aaa">/104</span>    |
| Lee          |                             97 |                                         | 522<span style="color: #aaa">/1335</span>    | 28<span style="color: #aaa">/112</span>    |

![](plots/covid-19-florida-icu-usage.png)

## Daily Testing

![](plots/covid-19-florida-tests-per-case.png)

<!-- ![](plots/covid-19-florida-change-new-cases.png) -->

![](plots/covid-19-florida-tests-percent-positive.png)

![](plots/covid-19-florida-test-and-case-growth.png)

## Cases and Deaths by Age

![](plots/covid-19-florida-weekly-events-by-age.png)

![](plots/covid-19-florida-age.png)

![](plots/covid-19-florida-age-deaths.png)

![](plots/covid-19-florida-age-sex.png)

## Sources

  - [Florida’s COVID-19 Data and Surveillance
    Dashboard](https://fdoh.maps.arcgis.com/apps/opsdashboard/index.html#/8d0de33f260d444c852a615dc7837c86)

  - [Florida Department of Health COVID-19 status
    page](http://www.floridahealth.gov/diseases-and-conditions/COVID-19/)

  - PDF Reports released daily on [Florida Disaster
    Covid-19](http://www.floridahealth.gov/diseases-and-conditions/COVID-19/)

The data structure and format of the released data changes frequently.
I’m keeping track of the impact of these changes in this repo in
[NEWS.md](NEWS.md).

## FL DOH Dashboard

One table is extracted from the [Florida’s COVID-19 Data and
Surveillance
Dashboard](https://fdoh.maps.arcgis.com/apps/opsdashboard/index.html#/8d0de33f260d444c852a615dc7837c86).

  - Current test counts:
    [data/covid-19-florida\_dash\_summary.csv](data/covid-19-florida_dash_summary.csv)

The `timestamp` column of indicates when the dashboard was polled for
changes.

Testing statistics prior to 2020-03-16 18:00:00 EDT were imported from
<https://covidtracking.com>.

![](screenshots/fodh_maps_arcgis_com__apps__opsdashboard.png)

## Snapshots and Data Capture

Initially this repo gathered data from the [Florida Department of Health
COVID-19 status
page](http://www.floridahealth.gov/diseases-and-conditions/COVID-19/).
This page has been [reformatted several
times](screenshots/floridahealth_gov__diseases-and-conditions__COVID-19.png)
and the data structure changes frequently, so currently I am collecting
snapshots of the webpage and not trying to parse these tables.

I am also capturing a snapshot (HTML) and screenshot of the [Florida’s
COVID-19 Data and Surveillance
Dashboard](https://fdoh.maps.arcgis.com/apps/opsdashboard/index.html#/8d0de33f260d444c852a615dc7837c86).
Data extracted from the dashboard are currently used for the current
test count summaries, available in
[covid-19-florida-tests.csv](covid-19-florida-tests.csv).

Prior to the afternoon of 2020-03-18, the dashboard reported
county-level case counts. Around 4pm on 2020-03-18, the dashboard layout
was modified, making these counts inaccessible. Similarly, prior to this
update I was relying on the reported “last update time” listed in the
dashboard, but since then I have been using the time at which the
dashboard was checked for any time stamps from this source.

On the same day that county-level data became inaccessible in the
dashboard, FL DOH started releasing PDF reports, which I have begun to
collect in [pdfs/](pdfs/). The name of the report contains a time stamp,
therefore I am periodically checking the FL DOH web page for the updated
link. I’ve been able to extract most of the data from the PDF tables.
Individual tables are stored in time stamped folders, using the time
stamp in the PDF file name. The unified data files are available in
[data/](data/) and prefixed with `covid-19-florida_pdf_`.


NEWS.md
# 2020-09-01

- I started to hit the strict 100MB file limit for the _Cases by Day_ data set
  archived from the FL DOH ArcGIS API. Instead of appending each new snapshot,
  only the latest snapshot will be kept, starting from 2020-08-31.

# 2020-07-17

- The ArcGIS feature layer with state summary statistics for the dashboard now
  include a state total where the `countyname` is `State`, or possibly `A State`.

# 2020-07-13

- Capture dataa from AHCA: hospitalizations, hospital bed capacity and icu capacity

# 2020-07-12

- I added the "deaths by day" to the data collected from the ArcGIS feature server.
  Note that the deaths included in this data are only Florida residents, and
  include residents who were diagnosed/isolated outside of Florida.
  
  https://services1.arcgis.com/CY1LXxl9zlJeBuRZ/ArcGIS/rest/services/Florida_COVID_19_Deaths_by_Day/FeatureServer/
  
- The health metrics data snapshot is only saved if the data has changed since
  the last update.

# 2020-06-23

- Now that the line list has passed 100k cases, the ArcGIS API requires scientific
  notation for result offests above 1e5. I've updated the query paging to fix this
  issue, but it seems that as of 2020-06-23 13:30 the line list has not yet been
  updated today (the number of rows is equal to yesterday's cases).

# 2020-06-08

- The location and file name format of the PDF reports was changed on 2020-06-04.
  I noticed on 2020-06-08 and updated my scripts, but the reports from
  2020-06-04 through 2020-06-07 were not stored.
  
- Reports on serology and antibody testing were added, presumably at the same
  time as the change in report location and naming (guess: 2020-06-04).

# 2020-05-21

- I updated the PDF extraction scripts to fix the extraction of the lab testing
  summary tables. In working on this, I noticed that the column headers have
  changed three times since these reports have been published:
  
  - Originally "Laboratory", "Negative", "Positive", "Total", "% Positive"
  - On 2020-03-18, "Laboratory" was changed to "Reporting Lab"
  - On 2020-04-11 (afternoon update), the "Inconclusive" column was added

# 2020-05-19

- I've found the updated link to the cases by zip code data and have resumed
  collecting those data.

# 2020-05-16

- The total tests reported on the dashboard between 2020-05-15 19:00 and 
  2020-05-16 08:00 EDT were out of range and are clearly erroneous.
  
- The data updates seem to be happening later in the day after 1pm, so I've
  gone back to using the last snapshot of the day for the day's counts.

# 2020-05-10

- The line list API was removed on 2020-05-09 from the public FeatureServer but 
  is still available under a new URL. I've updated my scripts to pull from that 
  URL and I don't think I've missed any updates.
  

# 2020-05-08

- The `event_date` column was removed on 2020-05-07 from the line list and
  as of 2020-05-08 the `age` column is missing for all cases. It's likely that
  the removal of the `event_date` is related to recent news coverage about
  early cases of coronavirus, see [issue #17](https://github.com/gadenbuie/covid19-florida/issues/17)
  for more discussion. I may have missed a day of updates (2020-05-07).
  
- The URL for the cases by day timeseries API changed, I may have also missed
  an update here.
  
# 2020-05-06 -- 2020-05-07

- It appears that there were issues with updates to the dashboard and data on
  these days. I believe the 11am update on 2020-05-07 was late (updated
  sometime between 12pm and 7pm). I've corrected the daily counts in the data
  used in my charts by treating the 2020-05-07 19:00 snapshot as if it happened
  at 2020-05-07 12:59.

# 2020-04-29

- Added **Health Metrics** data set that was released on 2020-04-28.
  <https://fdoh.maps.arcgis.com/home/item.html?id=492f02b473df4c8f8aa68fedf420c933>
  
- Fixed directory names in `pdfs/` so that they do not include `":"` (#18).

# 2020-04-25

- Starting Saturday April 25, the COVID-19 case and testing counts are only
  updated once daily, at around 11am. I've modified my summary plots and scripts
  to use the last snapshot of reported counts prior to 1pm, so that daily changes
  prior to the 2020-04-25 change are comparable to changes obsevered after.

# 2020-04-22 and 2020-04-23

- The evening update on 2020-04-22 may have come later than usual and my 
  snapshot may have missed the last update of the day. I've manually adjusted
  the `dash_summary` and `arcgis_summary` tables to match.
  
- As of 2020-04-23 in the morning, the line list is currently not available.
  This is causing the dashboard to show "0 Total Cases". I adjusted the snapshot
  scripts so that updates will be stored only when the table is available.

# 2020-04-13

- At some point in the last few days, the Cases variable in the Line List table
  served via ArcGIS changed from a numeric timestamp to a character `d/m/Y HH:MM`
  timestamp.
  
- Similarly Florida resident deaths were previously reported in the variable
  `fl_res_deaths`, which was renamed `c_fl_res_deaths` in the county-level
  summary table served via ArcGIS.

# 2020-04-09

- Additional tables that have been added to the dashboard are now also available
  via the arcgis API. I've added the Cases By Zip and Cases by Day tables to the
  data sets. The raw `geojson` for cases by zip are also included in
  `data/geojson`.
  
- I fixed an issue with the line list query to better utilize the arcgis API.
  I mised one or possibly two updates to this table in since 2020-04-08.

# 2020-03-26

- The total number of positive cases (regardless of Florida residency status)
  was remove from the dashboard in the evening. The dashboard now only shows the
  number of positive cases among Florida residents.

# 2020-03-25

- The division of ages in the case count by age range summary on Page 2 of the
  PDF report has changed from 0-9, 10-19, etc. to 0-4, 5-14, 15-24, etc. in the
  2020-03-25-1007 report.
  

R/process_pdf.R
library(readr)
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(lubridate)
library(glue)
library(fs)

fl_counties_w_spaces <- c("St. Johns", "Palm Beach", "Santa Rosa", "Indian River", "St. Lucie")

walk(
  dir_ls(here::here("R"), regexp = "process_pdf_.+R$"),
  sys.source, 
  env = globalenv()
)

find_cell <- function(table, pattern) {
  table %>% 
    set_names(seq_along(colnames(.))) %>% 
    mutate(row = row_number()) %>% 
    pivot_longer(cols = -row, names_to = "col", values_to = "cell") %>% 
    mutate_at(1:2, as.numeric) %>% 
    filter(str_detect(cell, pattern))
}

names_from_row <- function(table, row = 1) {
  colnames(table) <- unlist(table[row, ])
  table %>% slice(-row)
}

drop_empty <- function(table) {
  table %>% 
    mutate_if(is.character, ~ if_else(.x == "", NA_character_, .x)) %>% 
    janitor::remove_empty("rows") %>% 
    janitor::remove_empty("cols")
}

collapse_header_rows <- function(table, rows) {
  header <- slice(table, rows) %>% map_chr(paste, collapse = " ")
  table %>% 
    slice(-!!rows) %>% 
    bind_rows(header, .) %>% 
    names_from_row(1)
}

replace_with_nbsp <- function(x, strings) {
  for (string in strings) {
    replacement <- str_replace_all(string, " ", "\u00a0")
    x <- gsub(string, replacement, x, fixed = TRUE)
  }
  x
}

remove_nbsp <- function(x) {
  gsub("\u00a0", " ", x, fixed = TRUE)
}

extract_table <- function(
  table,
  top_left,
  top_right,
  bottom_left,
  numeric_cols = NULL,
  names_row = 1
) {
  col_starts <- find_cell(table, top_left)
  col_ends <- find_cell(table, top_right) %>% 
    filter(row == col_starts$row, col > col_starts$col)
  row_ends <- find_cell(table, bottom_left) %>% 
    filter(col == col_starts$col)
  
  x <- table %>%
    slice(col_starts$row:row_ends$row) %>% 
    select(col_starts$col:col_ends$col) %>% 
    names_from_row() %>% 
    drop_empty() %>% 
    janitor::clean_names()
  
  if (is.null(numeric_cols)) {
    x %>% 
      mutate_at(numeric_cols, as.numeric)
  } else x
}

read_table_pages <- function(
  pages,
  col_names, 
  pattern_filter = "\\d\\s{4,}\\d", 
  pattern_tab = "\\s{4,}", 
  replace_tab = "\t"
) {
  pages %>%
    str_split("\n") %>%
    map(str_subset, pattern = pattern_filter) %>%
    keep(~ length(.x) > 0) %>%
    map(str_replace_all, pattern = pattern_tab, replace_tab) %>%
    map_chr(paste, collapse = "\n") %>%
    quietly(readr::read_tsv, col_names = col_names)
}

extract_table_lines <- function(page_text, pattern_filter = "\\d\\s{4,}\\d") {
  page_text %>%
    str_split("\n") %>%
    map(str_subset, pattern = pattern_filter) %>%
    keep(~ length(.x) > 0)
}

add_total <- function(df, from) {
  if ("total" %in% names(df)) {
    if (!all(is.na(df[["total"]]))) {
      return(df)
    }
  }
  from <- intersect(from, names(df))
  df$total <- 0
  for (col in from) {
    df$total <- df$total + ifelse(is.na(df[[col]]), 0, df[[col]])
  }
  df
}

fix_missing_percent <- function(df, col = "percent") {
  idx_pct <- which(colnames(df) == col)
  idx_after <- seq_along(colnames(df))[-1:-idx_pct]
  df[!str_detect(df[[col]], "%"), idx_after] <- 
    df[!str_detect(df[[col]], "%"), idx_after - 1]
  df[!str_detect(df[[col]], "%"), idx_pct] <- NA_character_
  df
}

add_timestamp <- function(data, timestamp) {
  data %>% 
    mutate(timestamp = !!timestamp) %>% 
    select(timestamp, everything())
}

try_safely <- function(.f, ...) {
  .g <- purrr::safely(.f)
  res <- .g(...)
  if (is.null(res$error)) {
    return(res$result)
  } else {
    rlang::warn(res$error$message)
    return(NULL)
  }
}

quietly <- function(.x, .f, ...) {
  if (isTRUE(getOption("process_pdf_quietly", TRUE))) {
    suppressWarnings(.f(.x, ...))
  } else {
    .f(.x, ...)
  }
}

str_detect_pct <- function(x, pattern, exclude_empty = TRUE) {
  n_out_of <- if (exclude_empty) {
    sum(str_detect(x, "^\\s*$", negate = TRUE))
  } else {
    length(x)
  }
  sum(str_detect(x, pattern)) / n_out_of
}

process_pdf <- function(pdf_file) {
  out <- list(message = c())
  add_message <- function(...) {
    out$message <<- c(out$message, paste0(..., collapse = ""))
  }
  
  page_text <- if (grepl("xz$", pdf_file)) {
    pdftools::pdf_text(xzfile(pdf_file))
  } else {
    pdftools::pdf_text(pdf_file)
  }
  
  # Page 1 ------------------------------------------------------------------
  page_one_text <- page_text[1]
  
  timestamp_pdf <- get_timestamp_from_pdf_path(pdf_file)
  
  out$timestamp_pdf <- timestamp_pdf
  
  add_this_timestamp <- partial(add_timestamp, timestamp = timestamp_pdf)
  
  out$overall_counts <-
    page_one_text %>%
    str_extract_all("(Total (identified|tested)|Tested positive|Tested negative|Currently awaiting testing|Positive|Negative|Inconclusive|Awaiting BPHL.+?testing)\\s*[\\d,]+") %>%
    `[[`(1) %>% 
    str_trim() %>% 
    tibble(raw = .) %>% 
    separate(raw, c("group", "count"), sep = "\\s{3,}") %>% 
    mutate_at(vars(group), tolower) %>% 
    mutate_at(vars(count), str_remove_all, pattern = "\\s*,\\s*") %>% 
    mutate_at(vars(count), as.numeric) %>% 
    mutate(
      group = case_when(
        str_detect(group, "total (identified|tested)") ~ "total",
        str_detect(group, "positive") ~ "positive",
        str_detect(group, "negative") ~ "negative",
        str_detect(group, "awaiting|pending") ~ "pending",
        str_detect(group, "inconclusive") ~ "inconclusive",
        TRUE ~ str_replace_all(group, " ", "_")
      )
    ) %>% 
    pivot_wider(names_from = group, values_from = count) %>% 
    add_this_timestamp()
  
  if (!nrow(out$overall_counts)) {
    out$overall_counts <- tibble(timestamp = timestamp_pdf)
  }
  
  
  # Page 2 ------------------------------------------------------------------
  out_page_2 <- try_safely(process_cases_page_2, page_text, timestamp_pdf)
  if (is.null(out_page_2)) {
    add_message("Extraction failed for case summary tables from page 2")
  }
  
  out$cases_county <- out_page_2$cases_county
  out$cases_sex    <- out_page_2$cases_sex
  out$cases_age    <- out_page_2$cases_age
  
  # Testing by County --------------------------------------------------------
  out$county_testing <- try_safely(process_county_testing, page_text, timestamp_pdf)
  if (is.null(out$county_testing)) {
    add_message("County testing extraction failed for ", pdf_file)
  }
  
  # Community Spread ---------------------------------------------------------
  community_spread <- try_safely(process_community_spread, page_text, timestamp_pdf)
  if (!is.null(community_spread)) {
    add_message("This report includes community spread tables")
    out$community_spread_city <- community_spread$city
    out$community_spread_county <- community_spread$county
  }
  
  # Testing by Lab ----------------------------------------------------------
  pages_lab_testing <- page_text %>% 
    map_lgl(str_detect, pattern = "testing by lab") %>% 
    which()
  
  out$lab_testing <- try_safely(function() {
    col_header <- page_text[pages_lab_testing][[1]] %>% 
      str_extract("\n(Reporting lab|Laboratory).*\n.*\n") %>%
      str_trim() %>%
      str_split("\\s{3,}") %>% 
      .[[1]]
    
    col_names_std <- c(
      "Laboratory" = "reporting_lab",
      "Reporting lab" = "reporting_lab",
      "Negative" = "negative",
      "Positive" = "positive",
      "Total" = "total",
      "positive" = "percent_positive",
      "Inconclusive" = "inconclusive"
    )
    
    col_names <- unname(col_names_std[col_header])
    
    if (any(is.na(col_names))) return(NULL)
    
    testing_data <- if (timestamp_pdf <= ymd_h("2020-04-11 12", tz = "America/New_York")) {
      page_text[pages_lab_testing] %>%
        read_table_pages(col_names = c("reporting_lab", "negative", "positive", "percent", "total"))
    } else {
      col_starts <- c(1, 87, 104, 116, 128, 138)
      col_ends <- c(86, 103, 115, 127, 137, 160)
      
      page_text[pages_lab_testing] %>% 
        map(extract_table_lines) %>%
        map(~ map(.x, paste, collapse = "\n")) %>% 
        map_chr(paste, collapse = "\n") %>% 
        paste(collapse = "\n") %>% 
        readr::read_fwf(
          col_types = cols(.default = col_character()),
          col_positions = readr::fwf_positions(
            col_starts,
            col_ends, 
            col_names = c("reporting_lab", "inconclusive", "negative", "positive", "percent", "total")
          )
        )
    }
    
    testing_data %>%
      select(-matches("percent|total")) %>%
      filter(reporting_lab != "Total") %>%
      mutate_at(vars(matches("negative|positive|inconclusive")), str_remove_all, pattern = ",") %>%
      mutate_at(vars(matches("negative|positive|inconclusive")), as.numeric) %>%
      add_total(from = c("negative", "positive", "inconclusive")) %>%
      add_this_timestamp()
  })
  
  if (is.null(out$lab_testing)) {
    add_message("Lab test counts were not extracted")
  }
  
  # Line List ----------------------------------------------------------------
  out$line_list <- try_safely(process_line_list, page_text, timestamp_pdf)
  
  if (is.null(out$line_list)) {
    add_message("Line list tables were not parsed correctly and are not included")
  }
  
  if (length(out$message)) {
    for (msg in out$message) message(msg)
  }
  
  qc_pdf_read(out)
}

get_timestamp_from_pdf_path <- function(pdf_file) {
  tibble(file = pdf_file) %>% 
    mutate(
      date_string = str_extract(file, "\\d[\\d._-]{5,}"),
      ts = ymd_hm(date_string, truncated = 2, tz = "America/New_York"),
      fallback = with_tz(as_datetime(mdy(date_string)), tz = "America/New_York"),
      ts = coalesce(ts, fallback + hours(12))
    ) %>%
    pull(ts) %>%
    strftime("%F %T %Z", tz = "America/New_York")
}

process_and_output_pdf <- function(pdf_files, update_old = FALSE) {
  for (pdf_file in pdf_files) {
    if (!file_exists(pdf_file)) next
    timestamp_pdf <- get_timestamp_from_pdf_path(pdf_file)
    outdir <- path("pdfs", str_replace_all(timestamp_pdf, " ", "_"))
    outdir <- str_replace_all(outdir, ":", "")
    if (!isTRUE(update_old) && dir_exists(outdir)) {
      next
    }
    message("Processing ", pdf_file, "...")
    tables <- process_pdf(pdf_file)
    dir_create(outdir)
    if ("timestamp_pdf" %in% names(tables)) {
      write_lines(tables[["timestamp_pdf"]], path(outdir, "README.md"), append = FALSE)
    }
    if ("message" %in% names(tables) && length(tables[["message"]])) {
      cat(
        "\n", paste("-", tables[["message"]]), 
        file = path(outdir, "README.md"), 
        append = TRUE,
        sep = "\n"
      )
    }
    for (name in setdiff(names(tables), c("timestamp_pdf", "message"))) {
      if (is.character(tables[[name]])) {
        cat(
          "\n", tables[[name]], 
          file = path(outdir, "README.md"), 
          append = TRUE,
          sep = "\n"
        )
      } else if (inherits(tables[[name]], "data.frame")) {
        write_csv(tables[[name]], path(outdir, name, ext = "csv"))
      }
    }
  }
}

harmonize_pdf_data <- function() {
  dirs <- dir_ls(here::here("pdfs"), type = "directory", regexp = "2020")
  csvs <- dir_ls(dirs, regexp = "csv$")
  outdir <- here::here("data")
  dir_create(outdir)
  data <- map(csvs, read_csv, col_types = cols(.default = col_character()))
  out <- list()
  for (df_name in names(data)) {
    name <- path_ext_remove(path_file(df_name))
    out[[name]] <- bind_rows(out[[name]], data[[df_name]])
  }
  iwalk(out, ~ write_csv(
    .x, 
    path(outdir, paste0("covid-19-florida_pdf_", gsub("-", "_", .y)), ext = "csv")))
}

process_all_pdfs <- function() {
  library(furrr)
  plan(multicore)
  pdf_files <- fs::dir_ls(here::here("pdfs"), regexp = "(daily|state_reports).+pdf([.]xz)?$")
  .null <- future_map(pdf_files, process_and_output_pdf, .progress = TRUE)
  harmonize_pdf_data()
}